src/service/grpcweb.ts

```diff
@@ -221,22 +237,185 @@ function generateJavaScript(fileDescriptor: FileDescriptorProto, exportMap: Expo
+    .dedent().printLn(`};`)
+             .printLn(`grpc.invoke(${method.serviceName}.${method.nameAsPascalCase}, {`)
+      .indent().printLn(`request: requestMessage,`)
+               .printLn(`host: this.serviceHost,`)
+               .printLn(`metadata: metadata,`)
+               .printLn(`onMessage: function (responseMessage) {`)
+        .indent().printLn(`listeners.data.forEach(function (callback) {`)
+          .indent().printLn(`callback(responseMessage);`)
+        .dedent().printLn(`});`)
+      .dedent().printLn(`},`)
+               .printLn(`onEnd: function () {`)
+        .indent().printLn(`listeners.end.forEach(function (callback) {`)
+          .indent().printLn(`callback();`)
+        .dedent().printLn(`});`)
+                 .printLn(`listeners.data = [];`)
+                 .printLn(`listeners.end = [];`)
+      .dedent().printLn(`}`)
+    .dedent().printLn(`});`)
+             .printLn(`return {`)
+      .indent().printLn(`on: function (eventType, callback) {`)
+        .indent().printLn(`listeners[eventType].push(callback);`)
+      .dedent().printLn(`}`)
+    .dedent().printLn(`};`)
+  .dedent().printLn(`};`);
+}
+
+function printBidirectionalStubMethod(
+  printer: CodePrinter,
+  method: RPCMethodDescriptor
+) {
+  printer
+           .printLn(`${method.serviceName}.prototype.${method.nameAsCamelCase} = function ${method.nameAsCamelCase}() {`)
+    .indent().printLn(`throw new Error("Client streaming is not currently supported");`)
+  .dedent().printLn(`}`);
+}
+function printClientStreamStubMethod(
+  printer: CodePrinter,
+  method: RPCMethodDescriptor
+) {
+  printer
+           .printLn(`${method.serviceName}.prototype.${method.nameAsCamelCase} = function ${method.nameAsCamelCase}() {`)
+    .indent().printLn(`throw new Error("Bi-directional streaming is not currently supported");`)
+  .dedent().printLn(`}`);
+}
+
+function printServiceStubTypes(methodPrinter: Printer, service: RPCDescriptor) {
+  const printer = new CodePrinter(0, methodPrinter);
+
+  printer
+           .printLn(`export class ${service.name}Client {`)
+    .indent().printLn(`readonly serviceHost: string;`)
+             .printLn(`constructor(serviceHost: string);`)
+
+  service.methods.forEach((method: RPCMethodDescriptor) => {
+    if (method.requestStream && method.responseStream) {
+      printBidirectionalStubMethodTypes(printer, method);
+    } else if (method.requestStream) {
+      printClientStreamStubMethodTypes(printer, method);
+    } else if (method.responseStream) {
+      printServerStreamStubMethodTypes(printer, method);
+    } else {
+      printUnaryStubMethodTypes(printer, method);
+    }
+  });
+  printer.dedent().printLn("}");
+}
+
+function printUnaryStubMethodTypes(
+  printer: CodePrinter,
+  method: RPCMethodDescriptor
+) {
+  printer
+             .printLn(`${method.nameAsCamelCase}(`)
+      .indent().printLn(`requestMessage: ${method.requestType},`)
+               .printLn(`metadata?: grpc.Metadata,`)
```

Comment:
If we are matching the `grpc.makeUnaryRequest()` method then we should provide overloads so the caller can optionally omit metadata; which would look something like this:

```
  printer
             .printLn(`${method.nameAsCamelCase}(`)
      .indent().printLn(`requestMessage: ${method.requestType},`)
               .printLn(`metadata: grpc.Metadata,`)
               .printLn(`callback: (error: any, responseMessage: ${method.responseType}|null) => void`)
    .dedent().printLn(`): void;`)
             .printLn(`${method.nameAsCamelCase}(`)
      .indent().printLn(`requestMessage: ${method.requestType},`)
               .printLn(`callback: (error: any, responseMessage: ${method.responseType}|null) => void`)
    .dedent().printLn(`): void;`);
```

Note, this will require a small addition to the JavaScript implementation:

```
  printer
             .printLn(`${method.serviceName}Client.prototype.${method.nameAsCamelCase} = function ${method.nameAsCamelCase}(`)
      .indent().printLn(`requestMessage,`)
               .printLn(`metadata,`)
               .printLn(`callback`)
    .dedent().printLn(`) {`)
      .indent().printLn(`if (arguments.length === 2) {`)
        .indent().printLn(`callback = arguments[1];`)
      .dedent().printLn("}")
               .printLn(`grpc.unary(${method.serviceName}.${method.nameAsPascalCase}, {`)
```

On this note, it doesn't appear that the callback is optional in `grpc.makeUnaryRequest()` - should it be optional here?